File: python/1-50/32. Longest Valid Parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack, res, s = [0], 0, ')' + s
        for i in xrange(1, len(s)):
        	if s[i] == ")" and s[stack[-1]] == "(":
        		stack.pop()
        		res = max(res, i - stack[-1])
        	else:
        		stack.append(i)
        return res

File: cpp/1-50/32. Longest Valid Parentheses.cpp
class Solution {
public:
	int longestValidParentheses(string s) {
		stack<int> stack;
		stack.push(-1);
		int res = 0;
		for (int i = 0; i < s.size(); i++) {
			int temp = stack.top();
			if (temp >= 0 && s[i] == ')' && s[temp] == '(') {
			    stack.pop();
				res = max(res, i - stack.top());
			}
			else
				stack.push(i);
		}
		return res;
	}
};